### ISSUE TITLE
InstaPy too slow

### ISSUE THREAD

**Issue by asoomar** [InstaPy/InstaPy#1669]
My intended use of the bot is to like and comment on the most recent posts of a set of usernames. However, Instapy is taking to long to do this, and I have tried several things to expedite the process. I reduced the sleep time by using `session.set_sleep_reduce(1)` Additionally, I have taken out some of the sleep calls within the interact_by_user function and within the functions that are called on in interact_by_user.

This made it a little quicker, but I feel like because I am only trying to engage on recent posts, it makes sense to only grab around the last 3 links from each profile. Therefore, I was wondering how I could only grab 3 links instead of the bunch that it is currently doing.

A snippet sample output from running this on one profile is 
`INFO [2018-03-17 13:59:34] [wildnatureexplorers]  Getting review_nature image list...`
`INFO [2018-03-17 14:00:00] [wildnatureexplorers]  Interaction begin...`
`INFO [2018-03-17 14:00:26] [wildnatureexplorers]  Post [1/1]`

As you can see, it takes about 30 seconds each for these two processes, which totals to about 1 minute, and I was wondering what I could do to make this finish in about 15-30 seconds total instead of 1 minute.

**Comment by daniel-ed** [comment]
From my point of view the speed of InstaPy is perfect for flying "under the radar". If you get faster you WILL draw attention of instagram and the risk of being blocked gets bigger. With InstaPy I am able to generate 40-80 Followers per day, which is perfect for the "feel" of an organic growth. I am running it non-stop for 1 month now and never had problems. It's perfect!

**Comment by DemOnJR** [comment]
it's not slow to just target the audience! ;)

**Comment by asoomar** [comment]
I totally understand, but I am in engagement rounds where I have to comment and like hundreds of accounts in under an hour or two. Manually, I am able to do this and not get banned, but its just easier with a bot you know! 
I just want the speed to be increased to about 15-30 seconds, so I can actually utilize the bot to its fullest extent!

**Comment by sionking** [comment]
Remove any sleep() function call.

**Comment by asoomar** [comment]
I have already done that, but that is not the part that is taking long. The issue, I feel, is that the bot gets a bunch of links of the posts for the usernames chosen, and I need only at max three of the most recent posts for each username. Therefore, I was wondering what could be done so that the bot only checks the last few posts of each username, instead of all of them?

**Comment by CharlesCCC** [comment]
`but I am in engagement rounds where I have to comment and like hundreds of accounts in under an hour or two. ` this sounds like you will get banned pretty quickly.  
May I ask the reason why you would like to do that ? even manually, it sounds crazy amount.

**Comment by asoomar** [comment]
It is, but IG puts posts on the explore page based on the engagement within the first hour or so. If it is a lot, then IG boosts it to more people's explore page, which gets your content to thousands and thousands of likes and hundreds of followers per day based on exposure

**Comment by therealyamashita** [comment]
@CharlesCCC, can you provide a perfect quickstart set-up for running the application 24hrs and avoid getting banned?

**Comment by Tjorriemorrie** [comment]
@renanlaudencia I would not recommend using it 24/7. It's easy for IG to aggregate by hour and detect bots. Rather create 3hour sprints within a 16hour window. The [schedule example](https://github.com/timgrossmann/InstaPy#schedule) is very good

**Comment by asoomar** [comment]
@Tjorriemorrie @CharlesCCC Would you know how to reduce the amount of posts the interact_by_users goes through, as I only need the last 3 posts from each account, not all of them.

**Comment by asoomar** [comment]
@daniel-ed I totally understand how it works perfect for running it in your case. I used to run my account that way manually, and got the same results, but with engagement groups, my engagement has been more explosive and so has the growth in amount of followers, which is why I am trying to use InstaPy in a way to help me auto comment on users in my group

**Comment by CharlesCCC** [comment]
@renanlaudencia @asoomar ,   @Tjorriemorrie already provided a really good example. as he mentioned 3 hour interval with 16 hour window will be idea.
I personally used a schedule task to schedue it to run every 2 hours, each hour with 100 likes. Haven't had issue with this setting. 

Just make sure you don't like more than 100 per hour and don't like more than 1000 per day. (Don't try, exceed this amount, you will get soft-banned, because I was.)

**Comment by DemOnJR** [comment]
i created random hour script activity from 07:00 to 20:00 he does his job alone (one cycle per day).
if you need leave a message ...

**Comment by asoomar** [comment]
@charlesCCC I appreciate that, but my intended use is different, as I am basically only running it for a max of 3 hours a day. With engagement groups, I have gotten 100+ followers per day. Right now, I am mainly focusing on speeding up the interact_by_users function so the program only takes in the last 3 links of each profile instead of all of the links. Would you know how to do this?

**Comment by CharlesCCC** [comment]
@asoomar what kind of engagement groups you are talking about ? I'm curious how that can get you 100+ followers per day ? 

As for speed things up, the only thing I can think of is reduce the sleep time which already been implemented by you :(

**Comment by asoomar** [comment]
@charlesCCC You can learn all about it by looking up “Engagement Group Instagram” on youtube. Its a better way for your content to go viral!

**Comment by CharlesCCC** [comment]
@asoomar that is Incredible.  So, You basically want to use Instapy to help with the super intense engagement for that group. Now whatever stuff you said previously all make sense to me now. Unfortunately, I personally don't know if there is any better way to configure instapy go fast than currently speed.   Maybe you can put some bounty request to get someone help you implement this feature.

**Comment by g-names** [comment]
@asoomar I am trying to solve the same issue for the same reason. Did you get anywhere with this?

**Comment by achalpandeyy** [comment]
Did anyone managed to implement this?

**Comment by asoomar** [comment]
@g-names @achalpandeyy I was not able to solve this issue, but I found the root cause. Its because the program calls on the xpath.contains() function about 2 to 3 times. Anytime this is called, it takes at least 15 seconds in itself, which is why it is taking so long to comment and like a posts from a single account. I was able to make it a little faster by deleting the function call when checking to see if the account is private, as I know that I am only using this on public accounts. Other than that, I’m unsure how to make it any faster.

**Comment by asoomar** [comment]
@g-names @achalpandeyy I have used this to engage on about 300 accounts in my IG engagement groups, and I typically run this right before I sleep and it takes about 5 hours to finish. However, it did get me 100+ followers per day when I did use it. I would suggest to build up to engaging on 300 accounts, as you may be soft banned pretty quickly

**Comment by achalpandeyy** [comment]
@asoomar your engagement group allows you 5 hours to complete a round?!

**Comment by CharlesCCC** [comment]
@asoomar How does the engagement group actually works ?

**Comment by asoomar** [comment]
@achalpandeyy I’m not using it in groups that have timings for specific rounds. This is just for DM groups that are in IG. I’m in 30+ groups in IG itself. Typically, they want you to engage as soon as possible, but mainly they just want you to engage before you post your own photos in the group for them to engage on

**Comment by asoomar** [comment]
@CharlesCCC There are mainly 2 different types. One is where you like and comment at your earliest convenience once someone posts their photo on their page. Typically, you would engage before you post your own photo. 
There’s also another type, where everyone posts a photo at the same time, and everyone in the group engages on everyone else’s within some time limit, which is usually about an hour or so. 

There’s variations to these groups, like some are “like only” groups and some are “comment only” groups. If you look it up a bit online, you can definitely find a lot of info out there about them

**Comment by Areahints** [comment]
@asoomar it seems like you're really targeting the viral stories and explore section of Instagram, I have an idea, I was wondering what would happen if you had 3 accounts with 5k followers each and you set instapy to follow and interact with the followers of all 3 accounts meaning `Account A` would comment, follow and like `Account B & C` and vice versa, wouldn't that increase the overall engagement rate across all 15000 followers? and get the image or video into the explore section faster?

for the real reason I'm here, I'm trying to understand what potency ratio actually does and if I need it and also I'm looking for anyone's config that can make 100 follows at a run, @CharlesCCC any pointers?

**Comment by asoomar** [comment]
@Areahints I feel that you would end up with a majority of the same followers across the 3 accounts. In essence, it would seem as though you have 15k collectively but you will most likely end up having the same 5k people following each of the accounts. This will happen because you are targeting the same people across different channels. I am just speculating that this would happen, so its definitely something interesting to try out! 

Potency ratio is the amount of followers you have divided by the amount of people you are following. I don't really use this because I am in engagement groups where it is expected to for everyone to engage with everyone else, so I have a set of usernames where in one run, it likes and comments on everyone's most recent pic. 

As I was growing, I also realized that doing follow/unfollow on any scale isn't the best technique because it leads you to have low quality followers. Feel free to DM me on instagram @wildnatureexplorers if you want to join some of my groups or for some more techniques!

**Comment by Areahints** [comment]
Ok @asoomar I've sent the dm, thanks for the quick response here!

**Comment by SirLucTheFirst** [comment]
@asoomar I've been tinkering to speed up the liking process so we can randomize a delay of 3-7 seconds like the FollowLiker functionality. Like you said, this is a specific functionality for liking the last post of a list of usernames within a set time period; the botspeed is fine for all the other functionalities. One thing I noticed is that checking an already liked image takes a lot of time, and that happens often enough in engagement rounds. I tried stripping the like function and it's components to a bare minimum but I broke too many references in the script doing that. Anyway, did you get any further with it?
